core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -0,0 +1,444 @@
+      case SamWorkspaceRoles.owner        => Set(terraBucketWriterRole)
+      case SamWorkspaceRoles.writer       => Set(terraBucketWriterRole)
+      case SamWorkspaceRoles.reader       => Set(terraBucketReaderRole)
+      case _                              => Set.empty[String]
+    }
+
+  def setupFastPassForUserInNewWorkspace(workspace: Workspace): ReadWriteAction[Unit] = {
+    if (!config.enabled) {
+      logger.debug(s"FastPass is disabled. Will not grant FastPass access to ${workspace.toWorkspaceName}")
+      return DBIO.successful()
+    }
+
+    DBIO.from(quotaAvailableForNewWorkspaceFastPassGrants(workspace)).flatMap { quotaAvailable =>
+      if (quotaAvailable) {
+        logger
+          .info(s"Adding FastPass access for ${ctx.userInfo.userEmail.value} in workspace ${workspace.toWorkspaceName}")
+        val expirationDate = DateTime.now(DateTimeZone.UTC).plus(config.grantPeriod.toMillis)
+        for {
+          roles <- DBIO
+            .from(samDAO.listUserRolesForResource(SamResourceTypeNames.workspace, workspace.workspaceId, ctx))
+          petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, workspace.googleProjectId))
+          userAndPet = UserAndPetEmails(WorkbenchEmail(ctx.userInfo.userEmail.value), petEmail)
+          _ <- setupProjectRoles(workspace, roles, userAndPet, expirationDate)
+          _ <- setupBucketRoles(workspace, roles, userAndPet, expirationDate)
+          _ <- DBIO
+            .from(openTelemetry.incrementCounter("fastpass-granted-user", tags = openTelemetryTags).unsafeToFuture())
+        } yield ()
+      } else {
+        logger.info(
+          s"Not enough IAM Policy Role Binding quota available to add FastPass access for ${ctx.userInfo.userEmail.value} in workspace ${workspace.toWorkspaceName}"
+        )
+        DBIO.successful()
+      }
+    }
+  }
+
+  def setupFastPassForUserInClonedWorkspace(parentWorkspace: Workspace,
+                                            childWorkspace: Workspace
+  ): ReadWriteAction[Unit] = {
+    if (!config.enabled) {
+      logger.debug(s"FastPass is disabled. Will not grant FastPass access to ${parentWorkspace.toWorkspaceName}")
+      return DBIO.successful()
+    }
+
+    DBIO.from(quotaAvailableForClonedWorkspaceFastPassGrants(parentWorkspace, childWorkspace)).flatMap {
+      quotaAvailable =>
+        if (quotaAvailable) {
+          logger.info(
+            s"Adding FastPass access for ${ctx.userInfo.userEmail} in workspace being cloned ${parentWorkspace.toWorkspaceName}"
+          )
+          val expirationDate = DateTime.now(DateTimeZone.UTC).plus(config.grantPeriod.toMillis)
+          for {
+            petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, childWorkspace.googleProjectId))
+            userAndPet = UserAndPetEmails(WorkbenchEmail(ctx.userInfo.userEmail.value), petEmail)
+            _ <- setupBucketRoles(parentWorkspace, Set(SamWorkspaceRoles.reader), userAndPet, expirationDate)
```

Comment:
Why is there no need to set up project roles on a cloned workspace?

Reply:
because the project roles only concern compute jobs and bigquery stuff -- all of that should be happening in the child project

Reply:
I'm sorry, I don't follow. Don't we want to grant the same permissions for a cloned workspace that we do for a newly created workspace?

Reply:
No because a cloned workspace gets access to the parent workspace's bucket to prevent data duplication, but it doesnt get access to the compute so the parent doesnt pay for that.